[docs] Filters (#2686) (#70)
OpenCTI-Platform/docs: docs/usage/filters-migration.md

- the flat structure didn't enable filters imbrication at several levels.

```
// old filters format
FrontendFilters: Map<string, { id: string | null, value: string }>

BackendFilters {
    key: string, // an enum that should be in the list of the available filter keys for given the entity type
    values: string[],
    operator: string,
    filterMode: 'and' | 'or',
}
```

Because changing filters format impacts almost everything in the platform, we decided to do a complete refactoring once and for all. This refactoring will enable almost all complex combinations and will be compatible with our long term vision of filters in OpenCTI.

## Content: what has been done

The new filter implementation bring major changes in the way filtering is done.

- We change the filters formats (see FilterGroup type above):
    - In the frontend, an operator and a mode are stored for each key,
    - The new format enables filters imbrication thanks to the new attribute 'filterGroups',
    - The keys are of type string (no more static list of enums).
    - The 'values' attribute can no longer contain null values (use the 'nil' operator instead).

> - We also renamed some filter keys, to be consistant with the entities schema definitions.

Reviewer: consistent

Author: done

- We implemented the handling of the different operators and modes in the backend.

> - We introduced a new operator: 'nil' and 'not_nil', enabling to test wether an attribute is empty or not.

Reviewer 2: ```suggestion
- We introduced a new operator: 'nil' and 'not_nil', enabling to test if an attribute is empty or not.
```
As a side note, if you prefer to keep wheter, you need to add the missing 'h'

## Warnings: what you should do to ensure a correct migration

> We wrote a migration to convert all the old stored filters (filters contained in streams, taxii collections, feeds, triggers, workspaces) in the new format.

Reviewer 2: ```suggestion
We wrote a migration to convert all filters created prior version 5.12 (filters contained in streams, taxii collections, feeds, triggers, workspaces) in the new format.
```

> But you might have things to change for your own bits of code (your own connectors, queries, python scripts...). For each filter you hardcoded, you should change the filters format in the new one.

Reviewer 2: ```suggestion
But you might have to change your own custom code (your own connectors, queries, python scripts...). For each filter that is hardcoded, you should change the filters format in the new one.
```

> But you might have things to change for your own bits of code (your own connectors, queries, python scripts...). For each filter you hardcoded, you should change the filters format in the new one.

Reviewer 2: Question: is it really a 'you should change' and not a 'you must change before updating the platform to 5.12' ? If hardcoded filter prio to v5.12 are kept in custom code then it's not going to work anymore right ?

Author: right it's a must :) i'm going to change this :)

> To convert an old filters (in the old backend format) in the new format:

Reviewer 2: ```suggestion
To convert filters prior to version 5.12 in the new format:
```

> - rename the key if it is a key that has been changed in the migration.

Reviewer 2: ```suggestion
- rename the key if it is a key that has been changed in the migration. (The exhaustive list is in the "List of filter keys that have been renamed" section)
```

]
```

### Examples of filters in new format with nil operator and key renaming

Below is an example of the new filters format for an old filters indicating:

(entity_type = Report) AND (label = No label OR label1)

```
const oldBackendFilter = [
    {
      key: 'labelledBy',
      values: [label1_id, null],
      operator: 'eq',
      filterMode: 'or',
    },
    {
      key: 'entity_type',
      values: ['Report'],
      operator: 'eq',
      filterMode: 'or',
    },
];

const newFilters = {
    mode: 'and',
    filters: [
      {
        key: 'objectLabel',
        values: [label1_id],
        operator: 'eq',
        mode: 'or',
      },
      {
        key: 'objectLabel',
        values: [],
        operator: 'nil',
        mode: 'or',
      },
      {
        key: 'entity_type',
        values: ['Report'],
        operator: 'eq',
        mode: 'or',
      }
    ],
    filterGroups: [],
};
```

## Url with old filters
When going to an url with filters in the old format, the user will be redirected to the page but all the settings stored in local (filters, ordering mode...) will be deleted. A warning message is displayed, indicating the url contains filters in the old format.

> If you want to share a correct url, you need to copy an url from an OpenCTI instance after 5.12.

Reviewer 2: ```suggestion
If you want to share an url that works after 5.12 migration, you need to copy an url from an OpenCTI instance after 5.12.
```
